st2common/st2common/router.py

```diff
@@ -0,0 +1,386 @@
+from st2common.util.jsonify import json_encode
+from st2common.util.http import parse_content_type_header
+
+
+LOG = logging.getLogger(__name__)
+
+
+def op_resolver(op_id):
+    module_name, func_name = op_id.split(':', 1)
+    __import__(module_name)
+    module = sys.modules[module_name]
+    return functools.reduce(getattr, func_name.split('.'), module)
+
+
+def abort(status_code=exc.HTTPInternalServerError.code, message='Unhandled exception'):
+    raise exc.status_map[status_code](message)
+
+
+def abort_unauthorized(msg=None):
+    raise exc.HTTPUnauthorized('Unauthorized - %s' % msg if msg else 'Unauthorized')
+
+
+def extend_with_default(validator_class):
+    validate_properties = validator_class.VALIDATORS["properties"]
+
+    def set_defaults(validator, properties, instance, schema):
+        for property, subschema in six.iteritems(properties):
+            if "default" in subschema:
+                instance.setdefault(property, subschema["default"])
+
+        for error in validate_properties(
+            validator, properties, instance, schema,
+        ):
+            yield error
+
+    return jsonschema.validators.extend(
+        validator_class, {"properties": set_defaults},
+    )
+
+
+def extend_with_additional_check(validator_class):
+    def set_additional_check(validator, properties, instance, schema):
+        ref = schema.get("x-additional-check")
+        func = op_resolver(ref)
+        for error in func(validator, properties, instance, schema):
+            yield error
+
+    return jsonschema.validators.extend(
+        validator_class, {"x-additional-check": set_additional_check},
+    )
+
+
+def extend_with_nullable(validator_class):
+    validate_type = validator_class.VALIDATORS["type"]
+
+    def set_type_draft4(validator, types, instance, schema):
+        is_nullable = schema.get("x-nullable", False)
+
+        if is_nullable and instance is None:
+            return
+
+        for error in validate_type(validator, types, instance, schema):
+            yield error
+
+    return jsonschema.validators.extend(
+        validator_class, {"type": set_type_draft4},
+    )
+
+
+CustomValidator = jsonschema.Draft4Validator
+CustomValidator = extend_with_nullable(CustomValidator)
+CustomValidator = extend_with_additional_check(CustomValidator)
+CustomValidator = extend_with_default(CustomValidator)
+
+
+class NotFoundException(Exception):
+    pass
+
+
+class Response(webob.Response):
+    def __init__(self, body=None, status=None, headerlist=None, app_iter=None, content_type=None,
+                 *args, **kwargs):
+        # Do some sanity checking, and turn json_body into an actual body
+        if app_iter is None and body is None and ('json_body' in kwargs or 'json' in kwargs):
+            if 'json_body' in kwargs:
+                json_body = kwargs.pop('json_body')
+            else:
+                json_body = kwargs.pop('json')
+            body = json_encode(json_body).encode('UTF-8')
```

Comment:
Do we still correctly handle "user-friendly" response JSON encoding?

IIRC, in the past, by default we encoded it friendly for computer consumption by default so it was more compact, but if ?debug flag was used, we returned it user-friendly formatted.